src/libNode/Node.h

```diff
@@ -563,7 +563,12 @@ class Node : public Executable {
   bool IsShardNode(const PubKey& pubKey);
   bool IsShardNode(const Peer& peerInfo);
 
-  uint32_t CalculateShardLeader(uint16_t lastBlockHash, uint32_t sizeOfShard);
+  uint32_t CalculateShardLeaderFromDequeOfNode(uint16_t lastBlockHash,
+                                               uint32_t sizeOfShard,
+                                               const DequeOfNode& shardMembers);
+  uint32_t CalculateShardLeaderFromShard(uint16_t lastBlockHash,
```

Comment:
if it is not too much effort, can make a template container and merge the two functions

Reply:
Tried that approach but it can be tricky due to the differences in data structure
```
vector<pair<PubKey, Peer>>
vector<std::tuple<PubKey, Peer, uint16_t>>
```

Reply:
I think `std::get` would work on both.